source/CustomSale.cpp

```diff
@@ -0,0 +1,287 @@
+						
+						if(kid.Size() < 2 + add)
+							continue;
+						if(value)
+						{
+							if(add)
+								relativeOutfitPrices[outfit] += kid.Value(2);
+							else
+								relativeOutfitPrices[outfit] = kid.Value(1);
+							if(child.Size() == 2 + add)
+								relativeOutfitPrices[outfit] /= outfit->Cost();
+						}
+						else if(offset)
+						{
+							if(add)
+								relativeOutfitOffsets[outfit] += kid.Value(2);
+							else
+								relativeOutfitOffsets[outfit] = kid.Value(1);
+							if(child.Size() == 2 + add)
+								relativeOutfitOffsets[outfit] /= outfit->Cost();
+						}
+					}
+				else if(grandChild.Size() >= 2)
+				{
+					const Outfit *outfit = outfits.Get(grandChild.Token(0));
+					relativeOutfitPrices[outfit] = grandChild.Value(1) / outfit->Cost();
+				}
+			}
+		}
+		else if(token == "outfitter")
+		{
+			for(const DataNode &grandChild : child)
+			{
+				bool value = (grandChild.Token(0) == "value");
+				bool offset = (grandChild.Token(0) == "offset");
+				if(value || offset)
+					for(const DataNode &kid : grandChild)
+					{
+						bool add = (kid.Token(0) == "add");
+						const Sale<Outfit> *item = items.Get(kid.Token(add));
+
+						if(kid.Size() < 2 + add)
+							continue;
+						if(value)
+						{
+							if(add)
+								relativePrices[item] += kid.Value(2);
+							else
+								relativePrices[item] = kid.Value(1);
+						}
+						else if(offset)
+						{
+							if(add)
+								relativeOffsets[item] += kid.Value(2);
+							else
+								relativeOffsets[item] = kid.Value(1);
+						}
+					}
+				else if(grandChild.Size() >= 2)
+				{
+					const Sale<Outfit> *item = items.Get(grandChild.Token(0));
+					relativePrices[item] = grandChild.Value(1);
+				}
+			}
+		}
+		else if(token == "hidden" || token == "import")
+		{
+			for(const auto& it : show)
+				if(token == it.second)
+					sellType = it.first;
+		}
+		else if(token == "visible")
+			sellType = CustomSale::SellType::VISIBLE;
+		else if(token == "location")
+		{
+			if(child.Size() == 1)
+				locationFilter.Load(child);
+			else if(child.Size() == 2)
+			{
+				source = GameData::Planets().Get(child.Token(1));
+				if(child.HasChildren())
+					child.PrintTrace("Warning: location filter ignored due to use of explicit planet:");
+			}
+			else
+				child.PrintTrace("Warning: use a location filter to choose from multiple planets:");
+		}
+		else if(token == "conditions")
+			toApply.Load(child);
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+	}
+	if(sellType == SellType::NONE)
+		sellType = SellType::VISIBLE;
+}
+
+
+
+// Can only add between CustomSales of same sellType.
+bool CustomSale::Add(const CustomSale &other)
+{
+	// sellType::NONE means a new CustomSale created with no data and default sellType
+	if(this->sellType == SellType::NONE)
+		this->sellType = other.sellType;
+	else if(other.sellType != this->sellType)
+		return false;
+	for(const auto& it : other.relativePrices)
```

Comment:
Not really sure what you mean here exactly, auto &&it as an alternative but what was the first thing? I don't see the difference between `for(const auto& it : other.relativePrices)` and `for(const auto &it : other.relativePrices)`
guess I'll just make it be auto &&it, I have seen both in the code and I'm not sure if there is a reason to go for one over the other

Reply:
Yeah, it's a stylistic nit. We have the convention of placing type modifiers like "is a pointer" (`*`) and "is a reference" (`&`)  adjacent to the variable name rather than the type declaration. Hence `const auto &it` is preferred over `const auto& it`.

As far as what `auto&&` means, the `&&` there is a "universal forwarding reference" that ensures the associated variable has the exact same type, const-ness, etc as the thing it is referring to.

Reply:
Oh right sorry my brain just couldn't see the difference between the two